testing/server.go

```diff
@@ -1117,3 +1128,111 @@ func (s *DockerServer) createNetwork(w http.ResponseWriter, r *http.Request) {
 	var c = struct{ ID string }{ID: network.ID}
 	json.NewEncoder(w).Encode(c)
 }
+
+func (s *DockerServer) listVolumes(w http.ResponseWriter, r *http.Request) {
+	s.volMut.RLock()
+	result := make([]docker.Volume, 0, len(s.volStore))
+	for _, volumeCounter := range s.volStore {
+		result = append(result, docker.Volume{
+			Name:       volumeCounter.volume.Name,
+			Driver:     volumeCounter.volume.Driver,
+			Mountpoint: volumeCounter.volume.Mountpoint,
+		})
+	}
+	s.volMut.RUnlock()
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(result)
+}
+
+func (s *DockerServer) createVolume(w http.ResponseWriter, r *http.Request) {
+	var data struct {
+		*docker.CreateVolumeOptions
+	}
+	defer r.Body.Close()
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	volume := &docker.Volume{
+		Name:   data.CreateVolumeOptions.Name,
+		Driver: data.CreateVolumeOptions.Driver,
+	}
+	// If the name is not specified, generate one.  Just using generateID for now
+	if len(volume.Name) == 0 {
+		volume.Name = s.generateID()
+	}
+	// If driver is not specified, use local
+	if len(volume.Driver) == 0 {
+		volume.Driver = "local"
+	}
+	// Mount point is a default one with name
+	volume.Mountpoint = "/var/lib/docker/volumes/" + volume.Name
+
+	// If the volume already exists, don't re-add it.
+	exists := false
+	s.volMut.RLock()
+	if s.volStore != nil {
+		for _, vol := range s.volStore {
+			if vol.volume.Name == volume.Name {
+				exists = true
+				break
+			}
+		}
+	} else {
+		// No volumes, create volStore
+		s.volStore = make(map[string]*volumeCounter)
+	}
```

Comment:
Maybe here you should call s.volMut.RUnlock(), and then inside the if condition do the following:

```go
if !exists {
    s.volMut.Lock()
    defer s.volMut.Unlock()
    if _, ok := s.volStore[volume.Name]; !ok {
        s.volStore[volume.Name] = &volumeCounter{volume: *volume, count: 0}
    }
}
```

You may alternatively just use a regular lock, as this is a server intended for testing purposes.